x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -260,6 +261,15 @@ private void determineStartingRange(ByteRange range, List<Range> pendingRanges,
         }
     }
 
+    private LongConsumer progressConsumer(long rangeStart) {
+        assert Thread.holdsLock(ranges);
+        if (rangeStart == complete) {
+            return this::updateCompletePointer;
+        } else {
+            return null;
+        }
```

Comment:
Mostly making sure I understand how this works. We don't update `complete` unless every byte before it is fetched. We also [adjust](https://github.com/elastic/elasticsearch/blob/7916e6a231db48e73b7cb31889ba25d814c49d6d/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java#L258) `targetRange.start` so that we can expect one of the ranges will resume fetching from `complete`. And that range is the only one that should update the `complete` field?

Reply:
Exactly.